File: packages/portable/src/numbers.ts
import { geminate, rendaku } from './characters.js';

export const DIGIT_KANJI_DEFAULT = '〇一二三四五六七八九';
export const DIGIT_KANJI_LEGAL = '〇壱弐参四五六七八九拾';
export const POWER_KANJI = '一十百千万   億   兆   京';

type NumberClass = Readonly<{ type: 'jd' | 'p' | 'ad'; value: number }>;

const NUMBER_CLASSES: readonly (readonly [string, NumberClass])[] = [
  ['〇零', { type: 'jd', value: 0 }], ['一壱', { type: 'jd', value: 1 }],
  ['二弐', { type: 'jd', value: 2 }], ['三参', { type: 'jd', value: 3 }],
  ['四', { type: 'jd', value: 4 }], ['五', { type: 'jd', value: 5 }],
  ['六', { type: 'jd', value: 6 }], ['七', { type: 'jd', value: 7 }],
  ['八', { type: 'jd', value: 8 }], ['九', { type: 'jd', value: 9 }],
  ['十拾', { type: 'p', value: 1 }], ['百', { type: 'p', value: 2 }],
  ['千', { type: 'p', value: 3 }], ['万', { type: 'p', value: 4 }],
  ['億', { type: 'p', value: 8 }], ['兆', { type: 'p', value: 12 }],
  ['京', { type: 'p', value: 16 }],
  ['0０', { type: 'ad', value: 0 }], ['1１', { type: 'ad', value: 1 }],
  ['2２', { type: 'ad', value: 2 }], ['3３', { type: 'ad', value: 3 }],
  ['4４', { type: 'ad', value: 4 }], ['5５', { type: 'ad', value: 5 }],
  ['6６', { type: 'ad', value: 6 }], ['7７', { type: 'ad', value: 7 }],
  ['8８', { type: 'ad', value: 8 }], ['9９', { type: 'ad', value: 9 }]
];

const CHARACTER_NUMBER_CLASS = new Map<string, NumberClass>();
for (const [characters, numberClass] of NUMBER_CLASSES) {
  for (const character of characters) CHARACTER_NUMBER_CLASS.set(character, numberClass);
}

const DIGIT_KANA: Readonly<Record<number, string>> = {
  0: 'れい', 1: 'いち', 2: 'に', 3: 'さん', 4: 'よん',
  5: 'ご', 6: 'ろく', 7: 'なな', 8: 'はち', 9: 'きゅう'
};
const POWER_KANA: Readonly<Record<number, string>> = {
  1: 'じゅう', 2: 'ひゃく', 3: 'せん', 4: 'まん',
  8: 'おく', 12: 'ちょう', 16: 'けい'
};

export class NotANumberError extends Error {
  constructor(readonly text: string, readonly reason: string) {
    super(`"${text}" is not a number: ${reason}`);
    this.name = 'NotANumberError';
  }
}

export function numberToKanji(
  number: number,
  options: {
    readonly digits?: string;
    readonly powers?: string;
    readonly '1sen'?: boolean;
  } = {}
): string {
  const digits = options.digits ?? DIGIT_KANJI_DEFAULT;
  const powers = options.powers ?? POWER_KANJI;
  if (!Number.isInteger(number) || number < 0) {
    throw new Error('Number must be a non-negative integer');
  }
  if (number === 0) return digits[0]!;

  let magnitude = 1;
  let magnitudeCharacter = '';
  for (let index = 0, power = 1; index < powers.length && power <= number; index++, power *= 10) {
    const character = powers[index]!;
    if (character !== ' ') {
      magnitude = power;
      magnitudeCharacter = character;
    }
  }
  if (magnitude === 1) return digits[number]!;

  const quotient = Math.floor(number / magnitude);
  const remainder = number % magnitude;
  const quotientText = quotient === 1 && magnitude <= (options['1sen'] ? 100 : 1000)
    ? ''
    : numberToKanji(quotient, { ...options, '1sen': true });
  const remainderText = remainder === 0 ? '' : numberToKanji(remainder, options);
  return quotientText + magnitudeCharacter + remainderText;
}

function parseNumberClasses(classes: readonly NumberClass[], start: number, end: number): number {
  let greatestPower = 0;
  let greatestIndex = -1;
  for (let index = start; index < end; index++) {
    const item = classes[index]!;
    if (item.type === 'p' && item.value > greatestPower) {
      greatestPower = item.value;
      greatestIndex = index;
    }
  }
  if (greatestIndex < 0) {
    let number = 0;
    for (let index = start; index < end; index++) number = number * 10 + classes[index]!.value;
    return number;
  }
  if (greatestIndex === start) {
    return 10 ** greatestPower
      + (start + 1 < end ? parseNumberClasses(classes, start + 1, end) : 0);
  }
  return parseNumberClasses(classes, start, greatestIndex) * 10 ** greatestPower
    + (greatestIndex + 1 < end ? parseNumberClasses(classes, greatestIndex + 1, end) : 0);
}

export function parseNumber(input: string): number {
  const classes: NumberClass[] = [];
  for (const character of input) {
    const numberClass = CHARACTER_NUMBER_CLASS.get(character);
    if (!numberClass) throw new NotANumberError(input, `Invalid character: ${character}`);
    classes.push(numberClass);
  }
  return parseNumberClasses(classes, 0, classes.length);
}

function joinNumberKana(
  previousType: NumberClass['type'] | null,
  previousValue: number | null,
  type: NumberClass['type'],
  value: number,
  prefix: string,
  suffix: string
): string {
  if (previousType === 'jd' && previousValue === 1 && type === 'p') {
    if (value === 3 || value === 12 || value === 16) prefix = geminate(prefix);
  } else if (previousType === 'jd' && previousValue === 3 && type === 'p') {
    if (value === 2 || value === 3) suffix = rendaku(suffix);
  } else if (previousType === 'jd' && previousValue === 6 && type === 'p') {
    if (value === 2) {
      prefix = geminate(prefix);
      suffix = rendaku(suffix, false, true);
    } else if (value === 16) prefix = geminate(prefix);
  } else if (previousType === 'jd' && previousValue === 8 && type === 'p') {
    if (value === 2) {
      prefix = geminate(prefix);
      suffix = rendaku(suffix, false, true);
    } else if (value === 3 || value === 12 || value === 16) prefix = geminate(prefix);
  } else if (previousType === 'p' && previousValue === 1 && type === 'p') {
    if (value === 12 || value === 16) prefix = geminate(prefix);
  } else if (previousType === 'p' && previousValue === 2 && type === 'p' && value === 16) {
    prefix = geminate(prefix);
  }
  return prefix + suffix;
}

function groupToKana(group: readonly NumberClass[]): string {
  let output = '';
  let previousType: NumberClass['type'] | null = null;
  let previousValue: number | null = null;
  for (const item of group) {
    const kana = item.type === 'jd' ? DIGIT_KANA[item.value]
      : item.type === 'p' ? POWER_KANA[item.value]
      : '';
    output = joinNumberKana(previousType, previousValue, item.type, item.value, output, kana ?? '');
    previousType = item.type;
    previousValue = item.value;
  }
  return output;
}

export function numberToKana(
  number: number,
  options: {
    readonly separator?: string | null;
    readonly kanjiMethod?: (number: number) => string;
  } = {}
): string | string[] {
  const separator = options.separator === undefined ? ' ' : options.separator;
  const kanji = (options.kanjiMethod ?? numberToKanji)(number);
  const groups: NumberClass[][] = [];
  let group: NumberClass[] = [];
  let previousType: NumberClass['type'] | null = null;
  let previousValue: number | null = null;

  for (const character of kanji) {
    const item = CHARACTER_NUMBER_CLASS.get(character);
    if (!item) continue;
    if (!previousType || (
      item.type === 'p'
      && (previousType === 'jd' || (previousType === 'p' && item.value > (previousValue ?? 0)))
    )) {
      group.push(item);
    } else {
      groups.push(group);
      group = [item];
    }
    previousType = item.type;
    previousValue = item.value;
  }
  if (group.length > 0) groups.push(group);
  const readings = groups.map(groupToKana);
  return separator === null ? readings : readings.join(separator);
}

File: packages/rust-kernel/src/numbers.rs
//! Japanese number conversion with the same grouping and reading rules as the
//! portable TypeScript analyzer.

use std::fmt;

use crate::characters::{geminate, rendaku};

pub const DIGIT_KANJI_DEFAULT: &str = "〇一二三四五六七八九";
pub const DIGIT_KANJI_LEGAL: &str = "〇壱弐参四五六七八九拾";
pub const POWER_KANJI: &str = "一十百千万   億   兆   京";

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
enum NumberType {
    JapaneseDigit,
    Power,
    ArabicDigit,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
struct NumberClass {
    kind: NumberType,
    value: u32,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct NumberToKanjiError;

impl fmt::Display for NumberToKanjiError {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        formatter.write_str("Number must be a non-negative integer")
    }
}

impl std::error::Error for NumberToKanjiError {}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct NotANumberError {
    pub text: Vec<u16>,
    pub reason: Vec<u16>,
}

impl NotANumberError {
    pub fn message_utf16(&self) -> Vec<u16> {
        let mut message: Vec<u16> = "\"".encode_utf16().collect();
        message.extend_from_slice(&self.text);
        message.extend("\" is not a number: ".encode_utf16());
        message.extend_from_slice(&self.reason);
        message
    }
}

impl fmt::Display for NotANumberError {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        formatter.write_str(&String::from_utf16_lossy(&self.message_utf16()))
    }
}

impl std::error::Error for NotANumberError {}

#[derive(Clone, Debug, Eq, PartialEq)]
pub enum NumberKana {
    Joined(Vec<u16>),
    Groups(Vec<Vec<u16>>),
}

fn number_class(unit: u16) -> Option<NumberClass> {
    let digit = |kind, value| Some(NumberClass { kind, value });
    match unit {
        0x3007 | 0x96f6 => digit(NumberType::JapaneseDigit, 0),
        0x4e00 | 0x58f1 => digit(NumberType::JapaneseDigit, 1),
        0x4e8c | 0x5f10 => digit(NumberType::JapaneseDigit, 2),
        0x4e09 | 0x53c2 => digit(NumberType::JapaneseDigit, 3),
        0x56db => digit(NumberType::JapaneseDigit, 4),
        0x4e94 => digit(NumberType::JapaneseDigit, 5),
        0x516d => digit(NumberType::JapaneseDigit, 6),
        0x4e03 => digit(NumberType::JapaneseDigit, 7),
        0x516b => digit(NumberType::JapaneseDigit, 8),
        0x4e5d => digit(NumberType::JapaneseDigit, 9),
        0x5341 | 0x62fe => digit(NumberType::Power, 1),
        0x767e => digit(NumberType::Power, 2),
        0x5343 => digit(NumberType::Power, 3),
        0x4e07 => digit(NumberType::Power, 4),
        0x5104 => digit(NumberType::Power, 8),
        0x5146 => digit(NumberType::Power, 12),
        0x4eac => digit(NumberType::Power, 16),
        0x30..=0x39 => digit(NumberType::ArabicDigit, u32::from(unit - 0x30)),
        0xff10..=0xff19 => digit(NumberType::ArabicDigit, u32::from(unit - 0xff10)),
        _ => None,
    }
}

pub fn number_to_kanji(number: f64) -> Result<Vec<u16>, NumberToKanjiError> {
    number_to_kanji_with(
        number,
        &DIGIT_KANJI_DEFAULT.encode_utf16().collect::<Vec<_>>(),
        &POWER_KANJI.encode_utf16().collect::<Vec<_>>(),
        false,
    )
}

pub fn number_to_kanji_with(
    number: f64,
    digits: &[u16],
    powers: &[u16],
    one_sen: bool,
) -> Result<Vec<u16>, NumberToKanjiError> {
    if !number.is_finite() || number < 0.0 || number.fract() != 0.0 {
        return Err(NumberToKanjiError);
    }
    number_to_kanji_inner(number, digits, powers, one_sen)
}

fn number_to_kanji_inner(
    number: f64,
    digits: &[u16],
    powers: &[u16],
    one_sen: bool,
) -> Result<Vec<u16>, NumberToKanjiError> {
    if number == 0.0 {
        return digits
            .first()
            .copied()
            .map(|unit| vec![unit])
            .ok_or(NumberToKanjiError);
    }

    let mut magnitude = 1.0;
    let mut magnitude_character = None;
    let mut power = 1.0;
    for character in powers {
        if power > number {
            break;
        }
        if *character != b' ' as u16 {
            magnitude = power;
            magnitude_character = Some(*character);
        }
        power *= 10.0;
    }
    if magnitude == 1.0 {
        return digits
            .get(number as usize)
            .copied()
            .map(|unit| vec![unit])
            .ok_or(NumberToKanjiError);
    }

    let quotient = (number / magnitude).floor();
    let remainder = number % magnitude;
    let mut output = if quotient == 1.0 && magnitude <= if one_sen { 100.0 } else { 1000.0 } {
        Vec::new()
    } else {
        number_to_kanji_inner(quotient, digits, powers, true)?
    };
    output.push(magnitude_character.ok_or(NumberToKanjiError)?);
    if remainder != 0.0 {
        output.extend(number_to_kanji_inner(remainder, digits, powers, one_sen)?);
    }
    Ok(output)
}

fn parse_number_classes(classes: &[NumberClass]) -> f64 {
    let Some((greatest_index, greatest_power)) = classes
        .iter()
        .enumerate()
        .filter(|(_, item)| item.kind == NumberType::Power)
        .fold(None, |greatest, (index, item)| match greatest {
            Some((_, power)) if power >= item.value => greatest,
            _ => Some((index, item.value)),
        })
    else {
        return classes
            .iter()
            .fold(0.0, |number, item| number * 10.0 + f64::from(item.value));
    };

    let power = 10_f64.powi(greatest_power as i32);
    if greatest_index == 0 {
        return power + parse_number_classes(&classes[1..]);
    }
    parse_number_classes(&classes[..greatest_index]) * power
        + parse_number_classes(&classes[greatest_index + 1..])
}

fn next_js_character(input: &[u16], offset: usize) -> usize {
    if (0xd800..=0xdbff).contains(&input[offset])
        && input
            .get(offset + 1)
            .is_some_and(|unit| (0xdc00..=0xdfff).contains(unit))
    {
        2
    } else {
        1
    }
}

pub fn parse_number(input: &[u16]) -> Result<f64, NotANumberError> {
    let mut classes = Vec::new();
    let mut offset = 0;
    while offset < input.len() {
        let width = next_js_character(input, offset);
        if width == 1
            && let Some(class) = number_class(input[offset])
        {
            classes.push(class);
            offset += 1;
            continue;
        }
        let character = &input[offset..offset + width];
        let mut reason: Vec<u16> = "Invalid character: ".encode_utf16().collect();
        reason.extend_from_slice(character);
        return Err(NotANumberError {
            text: input.to_vec(),
            reason,
        });
    }
    Ok(parse_number_classes(&classes))
}

fn digit_kana(value: u32) -> &'static str {
    match value {
        0 => "れい",
        1 => "いち",
        2 => "に",
        3 => "さん",
        4 => "よん",
        5 => "ご",
        6 => "ろく",
        7 => "なな",
        8 => "はち",
        9 => "きゅう",
        _ => "",
    }
}

fn power_kana(value: u32) -> &'static str {
    match value {
        1 => "じゅう",
        2 => "ひゃく",
        3 => "せん",
        4 => "まん",
        8 => "おく",
        12 => "ちょう",
        16 => "けい",
        _ => "",
    }
}

fn join_number_kana(
    previous: Option<NumberClass>,
    item: NumberClass,
    prefix: &mut Vec<u16>,
    suffix: &mut Vec<u16>,
) {
    if let Some(previous) = previous {
        match (previous.kind, previous.value, item.kind, item.value) {
            (NumberType::JapaneseDigit, 1, NumberType::Power, 3 | 12 | 16)
            | (NumberType::JapaneseDigit, 6, NumberType::Power, 16)
            | (NumberType::JapaneseDigit, 8, NumberType::Power, 3 | 12 | 16)
            | (NumberType::Power, 1, NumberType::Power, 12 | 16)
            | (NumberType::Power, 2, NumberType::Power, 16) => *prefix = geminate(prefix),
            (NumberType::JapaneseDigit, 3, NumberType::Power, 2 | 3) => {
                *suffix = rendaku(suffix, false)
            }
            (NumberType::JapaneseDigit, 6 | 8, NumberType::Power, 2) => {
                *prefix = geminate(prefix);
                *suffix = rendaku(suffix, true);
            }
            _ => {}
        }
    }
    prefix.append(suffix);
}

fn group_to_kana(group: &[NumberClass]) -> Vec<u16> {
    let mut output = Vec::new();
    let mut previous = None;
    for item in group {
        let kana = match item.kind {
            NumberType::JapaneseDigit => digit_kana(item.value),
            NumberType::Power => power_kana(item.value),
            NumberType::ArabicDigit => "",
        };
        let mut suffix: Vec<u16> = kana.encode_utf16().collect();
        join_number_kana(previous, *item, &mut output, &mut suffix);
        previous = Some(*item);
    }
    output
}

fn kana_groups(kanji: &[u16]) -> Vec<Vec<u16>> {
    let mut groups: Vec<Vec<NumberClass>> = Vec::new();
    let mut group = Vec::new();
    let mut previous = None;
    for character in kanji {
        let Some(item) = number_class(*character) else {
            continue;
        };
        let continues = previous.is_none_or(|previous: NumberClass| {
            item.kind == NumberType::Power
                && (previous.kind == NumberType::JapaneseDigit
                    || (previous.kind == NumberType::Power && item.value > previous.value))
        });
        if !continues {
            groups.push(std::mem::take(&mut group));
        }
        group.push(item);
        previous = Some(item);
    }
    if !group.is_empty() {
        groups.push(group);
    }
    groups.iter().map(|items| group_to_kana(items)).collect()
}

pub fn number_to_kana(number: f64) -> Result<Vec<u16>, NumberToKanjiError> {
    match number_to_kana_with_separator(number, Some(&[b' ' as u16]))? {
        NumberKana::Joined(value) => Ok(value),
        NumberKana::Groups(groups) => {
            Ok(groups
                .into_iter()
                .enumerate()
                .fold(Vec::new(), |mut joined, (index, group)| {
                    if index > 0 {
                        joined.push(b' ' as u16);
                    }
                    joined.extend(group);
                    joined
                }))
        }
    }
}

pub fn number_to_kana_with_separator(
    number: f64,
    separator: Option<&[u16]>,
) -> Result<NumberKana, NumberToKanjiError> {
    let kanji = number_to_kanji(number)?;
    Ok(number_to_kana_from_kanji(&kanji, separator))
}

pub fn number_to_kana_from_kanji(kanji: &[u16], separator: Option<&[u16]>) -> NumberKana {
    let readings = kana_groups(kanji);
    let Some(separator) = separator else {
        return NumberKana::Groups(readings);
    };
    let mut output = Vec::new();
    for (index, reading) in readings.iter().enumerate() {
        if index > 0 {
            output.extend_from_slice(separator);
        }
        output.extend_from_slice(reading);
    }
    NumberKana::Joined(output)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn u(value: &str) -> Vec<u16> {
        value.encode_utf16().collect()
    }

    #[test]
    fn preserves_exact_error_text_in_utf16() {
        let error = parse_number(&[0x4e00, 0xd83d, 0xde00]).unwrap_err();
        let mut expected = u("\"一");
        expected.extend([0xd83d, 0xde00]);
        expected.extend(u("\" is not a number: Invalid character: "));
        expected.extend([0xd83d, 0xde00]);
        assert_eq!(error.message_utf16(), expected);
        assert_eq!(
            NumberToKanjiError.to_string(),
            "Number must be a non-negative integer"
        );
        assert_eq!(
            number_to_kanji_with(
                0.0,
                &[0xd83d],
                &POWER_KANJI.encode_utf16().collect::<Vec<_>>(),
                false,
            )
            .unwrap(),
            vec![0xd83d],
        );
    }
}
